pkgs/applications/audio/sonobus/default.nix

```diff
@@ -0,0 +1,83 @@
+{ lib
+, pkg-config
+, stdenv
+, fetchFromGitHub
+, alsa-lib
+, cmake
+, fetchzip
+, freetype
+, libGL
+, libX11
+, libXcursor
+, libXext
+, libXinerama
+, libXrandr
+, libjack2
+, libopus
+, opusTools
+, curl
+, gtk3
+, webkitgtk
+}:
+
+stdenv.mkDerivation rec {
+  pname = "sonobus";
+  version = "1.6.0";
+
+  src = fetchFromGitHub {
+    owner = "sonosaurus";
+    repo = "sonobus";
+    rev = version;
+    sha256 = "sha256-Qz5nUgT/p6BuO1VGz5PiDxvv/1ctBOTVrG4BAbJS9/g=";
+    fetchSubmodules = true;
+  };
+
+  nativeBuildInputs = [
+    cmake
+    pkg-config
+  ];
+
+  buildInputs = [
+    alsa-lib
+    freetype
+    libGL
+    libX11
+    libXcursor
+    libXext
+    libXinerama
+    libXrandr
+    libjack2
+    libopus
+    opusTools
+    curl
+    gtk3
+    webkitgtk
+  ];
+
+  # To add the libraries to the rpath, needed or the program segfaults
+  NIX_LDFLAGS = "-rpath ${lib.makeLibraryPath buildInputs}";
+  dontPatchELF = true;
```

Comment:
I'm not sure if turning off patchelf has any significant side-effects. I generally see packages solve this by add a `wrapProgram` (from `makeWrapper` setup hook) in `installPhase` or `postInstall` to prefix `LD_LIBRARY_PATH`, but I've no strong opinions since I do see this construction elsewhere

The `libGL` and `libX*` deps could be moved to such a call (since they aren't needed at build-time) if you ended up preferring that:

```
    wrapProgram $out/bin/sonobus \
      --prefix LD_LIBRARY_PATH : "${lib.makeLibraryPath [ libGL libX11 libXcursor libXext libXinerama libXrandr ]}"
```

Reply:
Well as I understand rpath is supposed to be cleaner than `LD_LIBRARY_LOAD` no? (e.g. it works with setuid binaries, don't need wrapper…). It's just that by default nix over-optimizes the rpath for some applications.

Reply:
Yeah but `dontPatchELF` prevents it from removing a lot of unnecessary runtime references. See the difference in closure size below:

```
[nix-shell:~/.cache/nixpkgs-review/pr-211341/nixpkgs]$ nix path-info -Sh /nix/store/p32vnnl6pw72zs53m9c3w0qpbbpk5q1q-sonobus-1.6.2  # straight from this PR as of d0ad888a4
/nix/store/p32vnnl6pw72zs53m9c3w0qpbbpk5q1q-sonobus-1.6.2	 593.3M

[nix-shell:~/.cache/nixpkgs-review/pr-211341/nixpkgs]$ nix path-info -Sh /nix/store/c1w40v5wfkva9l5b1vi2i9q3ybbxrqir-sonobus-1.6.2  # with my suggested change
/nix/store/c1w40v5wfkva9l5b1vi2i9q3ybbxrqir-sonobus-1.6.2	  97.2M
```

If we want to set rpath manually (e.g. if it does turn out to be needed for the vst plugins, which I didn't test), I would add it in a `postFixup` like `patchelf --add-rpath ${lib.makeLibraryPath [ libGL libX11 libXcursor libXext libXinerama libXrandr ]} $out/bin/sonobus $out/lib/vst3/*/Contents/x86_64-linux/*.so` and remove them from `buildInputs` since the cross splicing would still resolve the target deps correctly

Reply:
Yeah I am fine if it just does have to be an rpath over all of the inputs (the amount I actually care is probably less than I made it sound), but if we want to do this optimization, it could be convenient for some

I'm preparing a diff in a minute to explain exactly how to patch in an additional rpath to stuff

Reply:
Ok thanks. Just, my reason to prefer `NIX_LDFLAGS = "-rpath…"` over manually calling `patchelf` is that it is easy with the last method to forget some files to patch (especially if the library is getting new plugins format later).

Reply:
This is the diff. I keep forgetting that `autoPatchelfHook` automates this so it's actually pretty simple

```diff
diff --git a/pkgs/applications/audio/sonobus/default.nix b/pkgs/applications/audio/sonobus/default.nix
index 26c6a0f0504..da82fcc5a86 100644
--- a/pkgs/applications/audio/sonobus/default.nix
+++ b/pkgs/applications/audio/sonobus/default.nix
@@ -2,6 +2,7 @@
 , pkg-config
 , stdenv
 , fetchFromGitHub
+, autoPatchelfHook
 , alsa-lib
 , cmake
 , freetype
@@ -31,6 +32,7 @@ stdenv.mkDerivation rec {
   };
 
   nativeBuildInputs = [
+    autoPatchelfHook
     cmake
     pkg-config
   ];
@@ -38,12 +40,6 @@ stdenv.mkDerivation rec {
   buildInputs = [
     alsa-lib
     freetype
-    libGL
-    libX11
-    libXcursor
-    libXext
-    libXinerama
-    libXrandr
     libjack2
     libopus
     curl
@@ -51,9 +47,14 @@ stdenv.mkDerivation rec {
     webkitgtk
   ];
 
-  # To add the libraries to the rpath, needed or the program segfaults
-  NIX_LDFLAGS = "-rpath ${lib.makeLibraryPath buildInputs}";
-  dontPatchELF = true;
+  runtimeDependencies = [
+    libGL
+    libX11
+    libXcursor
+    libXext
+    libXinerama
+    libXrandr
+  ];
 
   postPatch = lib.optionalString (stdenv.isLinux) ''
     # needs special setup on Linux, dunno if it can work on Darwin
```

Reply:
Oh thanks, I thought autopatchelf would only work for pre-built binaries, good to know! Maybe the wrapper adds bash to the closure (unless you used the binary version).

Reply:
Oh yeah it probably was bash. I can agree with you that wrappers suck, though, so I am glad the autoPatchelf hook makes it very idiomatic to express and manages to catch all shared objects and binaries too